### DISCUSSION TITLE
checkInternetConnection ios problem

### DISCUSSION THREAD

**Discussion by avcr2019** [airsdk/Adobe-Runtime-Support#3319] [Q&A]
Welcome, everyone
I hope you all have a nice day
I have a somewhat strange problem and I don't know the cause yet. I have been trying to solve it for a week and have not found a solution yet.
The problem: The loss of Internet connection is not checked within the application, and the application remains convinced that there is an Internet connection, and thus it causes me losses in that it displays content that is supposed to appear in the presence of the Internet, but it shows it even if there is no Internet connection.
More clearly, it is assumed that there is a movie clip called (errordia) and it is located above the content that will appear with the Internet in the application. If there is Internet, the movie clip (errordia) will disappear, and if there is no Internet connection, the movie clip (errordia) will continue to appear above the content until it opens. Internet user.
I wrote more than one code, and each time the code works very successfully on the Android system and there are no problems. If there is Internet, the movie clip (errordia) disappears and the user can watch the content. If he closes the Internet, the movie clip (errordia)appears at the same moment above the content and there are no problems, but the problem is in the system. (IOS SYSTEM ) It is very difficult and caused me psychological distress for a week and I did not find a solution. Because it is not working properly. Whereas, if the user opens the application for the first time and there is an Internet connection, the movie clip (errordia) disappears correctly, and this is good. But when the user turns off the Internet while he is still inside the application, it is supposed that the movie clip (errordia) appears again because the Internet connection has been lost, but this does not happen while the movie clip continues to disappear as if the phone is connected to the Internet!!!!! The user watches paid content for free
I don't know how this is, knowing that the same code works correctly on the Android system
Knowing that the code works correctly only once on the (IOS) system, since if the user opens the application for the first time while he is not connected to the Internet, he will see the movie clip (errordia) above the content, and this is a beautiful thing, so it turns on the Internet so that the movie clip (errordia) disappears, and indeed it disappears when verified. There is Internet, but if the user turns off the Internet again, the movie clip (errordia) will continue to disappear, haha, and I do not know how, and this is a problem.
I will show you more than one example that I tried differently but with the same result. I hope to find a solution or a specific code to solve the problem

first code :
`import flash.net.URLLoader;
import flash.events.Event;
import flash.events.IOErrorEvent;
import flash.net.URLRequest;
import flash.events.MouseEvent;
import flash.display.MovieClip;

var loader:URLLoader = new URLLoader();
loader.addEventListener(Event.COMPLETE, onCompleteLoading);
loader.addEventListener(IOErrorEvent.IO_ERROR, onErrorLoading);
loader.load(new URLRequest("https://www.google.com"));

;

function onCompleteLoading(e:Event):void
{
	
	errordia.visible= false;	
}

function onErrorLoading(e:IOErrorEvent):void
{
	errordia.visible= true;
	}`

second code :
`import flash.utils.Timer;
import flash.events.TimerEvent;
import flash.net.URLLoader;
import flash.net.URLRequest;
import flash.events.Event;
import flash.events.IOErrorEvent;

var internetCheckTimer:Timer = new Timer(1000); 
internetCheckTimer.addEventListener(TimerEvent.TIMER, checkInternetConnection);
internetCheckTimer.start(); 

function checkInternetConnection(event:TimerEvent):void {
    var urlLoader:URLLoader = new URLLoader();
    urlLoader.addEventListener(Event.COMPLETE, onCompleteLoading);
    urlLoader.addEventListener(IOErrorEvent.IO_ERROR, handleIOError);
    urlLoader.load(new URLRequest("https://www.google.com"));
}

errordia.visible = true;

var isInternetAvailable:Boolean = false; 

function handleIOError(event:IOErrorEvent):void {
   
    isInternetAvailable = false;
    showHideElements(); 
}

function onCompleteLoading(e:Event):void {
   
    isInternetAvailable = true; 
    showHideElements(); 
}

function showHideElements():void {
    errordia.visible = !isInternetAvailable;

}

stage.addEventListener(Event.ACTIVATE, clearCache);

function clearCache(event:Event):void {
    internetCheckTimer.stop(); 
    internetCheckTimer.reset(); 
}`

Video to illustrate the problem :

https://github.com/airsdk/Adobe-Runtime-Support/assets/51138931/769213d4-2024-4de3-a19e-c2ebe4019725

**Comment by ghost** [comment]
You should use NetworkInfo ANE from Distriqt to check reachability.

**Reply by avcr2019** [reply]
> You should use NetworkInfo ANE from Distriqt to check reachability.

Thank you for the response, my dear friend.
But will this library implement what is required or will it cause problems with the (IOS) system? Because I searched the library and found some people complaining that there are problems with this library in the (IOS).
Is there another way without using any external libraries?

**Reply by avcr2019** [reply]
> You should use NetworkInfo ANE from Distriqt to check reachability.

I will buy this library. But I want to know whether it will do the job I need or not? I know that the library determines the type of network the phone is connected to, whether Wi-Fi or other, but I do not want that. I want to know if the user has turned off the Internet. I will execute certain codes, such as closing a screen, closing content, etc. Does this library do that, or does it only detect the type of connection?
Because when I went to see the codes for it, I did not find anything that would allow me to control the elements when the Internet is shut down, while I only know if there is a connection or not.
![Screenshot 2024-06-28 at 5 18 14 PM](https://github.com/airsdk/Adobe-Runtime-Support/assets/51138931/f139a8de-b677-48ce-b25a-e124223fe4b4)

**Reply by ghost** [reply]
To check connection just call NetworkInfo.networkInfo.isReachable(), works on iOS and Android.

**Comment by avcr2019** [comment]
Please, is there any solution to this problem? My content is downloaded in full without any profits

**Reply by ajwfrost** [reply]
Just regarding the original code you posted .. if you're getting a 'complete' event saying it's loaded that URLLoader content, even though you know it can't have done, can you try changing the URL each time?
e.g.
```
urlLoader.load(new URLRequest("https://www.google.com/search?q=test" + flash.utils.getTimer()));
```
So that every time it's called, it's a different URL, and there won't then be any caching..?

**Reply by avcr2019** [reply]
> urlLoader.load(new URLRequest("https://www.google.com/search?q=test" + flash.utils.getTimer()));

welcome my dear friend
I wish you a happy day
I tried this code of yours
Code
 Indeed, two things happened, one happy and the other sad
The happy news is that when I actually turned off the Internet on the phone, the content was successfully closed, but the sad news is that when I open the Internet again and the Internet is already connected successfully, the content does not open and I remain convinced that there is no Internet. It seems that the opposite of my problem happened. Thank you for your effort, interaction and help, I appreciate it. I hope you can help me complete the code correctly because it seems that this is the code that will solve the problem, but it needs minor modifications.
Wait your reply

**Reply by ajwfrost** [reply]
I can't really see anything wrong with your code; I would suggest adding a bunch of traces in there and remote-debugging to see what's actually going wrong - whether there's something that's caused the timer to stop, for example. I don't know of a reason for the url loader to fail when the internet comes back on again..

**Reply by avcr2019** [reply]
> I can't really see anything wrong with your code; I would suggest adding a bunch of traces in there and remote-debugging to see what's actually going wrong - whether there's something that's caused the timer to stop, for example. I don't know of a reason for the url loader to fail when the internet comes back on again..

Thank you for your response and interest
I apologize for the delay in responding
Unfortunately the code does not work. In other words, sometimes it works and sometimes it doesn't work, and I don't know why. I do not know how to track the matter in order to find out why it stopped working. I use Adobe Animate and I use the Harman package. Can you help me how to find out the tracking so that I can send it to you?
- I tried to write another code and it actually works, but it is not as efficient as the line that you wrote, so I will share it so that everyone can benefit. If someone faces the same problem, he can use the code. But I repeat my words, the code that you wrote is much better because it shows and hides the element very quickly, while What I wrote is not fast enough, but only temporarily until we solve the problem
 The code I wrote alternatively works now:
 `import flash.net.URLLoader;
import flash.events.Event;
import flash.events.IOErrorEvent;
import flash.net.URLRequest;
import flash.events.MouseEvent;
import flash.display.MovieClip;
import flash.utils.Timer;
import flash.events.TimerEvent;
import flash.events.Event;

var loader:URLLoader = new URLLoader();
loader.addEventListener(Event.COMPLETE, onCompleteLoading);
loader.addEventListener(IOErrorEvent.IO_ERROR, onErrorLoading);

var checkInternetTimer:Timer = new Timer(5000);
checkInternetTimer.addEventListener(TimerEvent.TIMER, checkInternetConnection);
checkInternetTimer.start();

function checkInternetConnection(event:TimerEvent):void {
    try {
        loader.load(new URLRequest("https://www.google.com?timestamp=" + new Date().time));
    } catch (error:Error) {
        handleError();
    }
}

function onCompleteLoading(e:Event):void {
    hideErrorDia();
}

function onErrorLoading(e:IOErrorEvent):void {
    handleError();
}

function handleError():void {
    showErrorDia();
    checkInternetTimer.stop();
    var retryTimer:Timer = new Timer(3000, 1); 
    retryTimer.addEventListener(TimerEvent.TIMER, function(event:TimerEvent):void {
        hideErrorDia();
        checkInternetTimer.start();
    });
    retryTimer.start();
}



function showErrorDia():void {
            errordia.visible = true;

}

function hideErrorDia():void {
     errordia.visible = false;

}`
